RemoveClassRefactoring: new generation

Ducasse opened the pull request:
Stupid bot just check the commit list please

Ducasse committed:
Refactor: remove choice logic from Syc Command

Ducasse committed:
Merge branch 'tryingToRescueRemoveClassWork' into 2023-09-16-RestartRemoveClass

balsa-sarenac committed:
Feat: expose remove method 2 command

balsa-sarenac committed:
Fix: add missing methods to enable first version of remove class driver

balsa-sarenac committed:
Merge branch '2023-09-16-RestartRemoveClass' into 2023-09-16-RestartRemoveClass

Ducasse committed:
Merge pull request #1 from balsa-sarenac/2023-09-16-RestartRemoveClass

Restart remove class refactoring

Ducasse committed:
pull up refactoring to RBDriver

Ducasse committed:
Refactor: remove `preconditionNoTopClasses:`

Ducasse committed:
Refactor: split preconditions on applicability and breaking changes

Ducasse committed:
Feat: introduce `prepareForExecution` hook

Ducasse committed:
Docs: comment explaining that we want to unify preconditions checking API

Ducasse committed:
Feat: custom condition for checking if a group of classes have references

Ducasse committed:
Refactor: remove precondition checking logic from `removeClassChanges`

This logic is moved to driver for now.

Ducasse committed:
Refactor: use `prepareForExecution` hook in RemoveClassRefactoring

Now preconditions don't need to create RBClasses but instead check their preconditions.

Ducasse committed:
Fix: name error for `ClassesHaveNoReferencesCondition`

Ducasse committed:
Refactor: use `ClassesHaveNoReferencesCondition` in RemoveClassDriver

Now driver can ask what are violators and give nice meaningful feedback.

Ducasse committed:
Fix: add missing method protocol

Ducasse committed:
Fix: composite refactorings never invoked `prepareForExecution`

Move `prepareForExecution` to `generateChanges` since composites
only invoke `generateChanges` and `execute` is invoked only when
scripting and some commands.

Ducasse committed:
Feat: introduce `prepareForInteractiveMode` Refactoring method

This method is responsible to:
- invoke prepare for execution so that all instance variables are
correctly set
- check applicability preconditions

Ducasse committed:
Refactor: remove side effect from `prepareForInteractiveMode`

Remove `applicabilityPreconditions` checking code from `preapreForInteractiveMode`.
Driver can now check the applicability preconditions after the refactoring is configured.
With this new design we've removed the need for error handling code from RemoveClassDriver!

Ducasse committed:
Refactor: remove class symbol users precondition from breaking changes

We still leave this method so users can check users of a class if they want,
but this will not be the default behavior any more.

Ducasse committed:
Fix: add protocol to `preconditionHasNoReferences`

Ducasse committed:
Refactor: rename temp to reflect condition name

Ducasse committed:
Before the final UI.

Ducasse committed:
Removing some unused method and bals will fix the rest

balsa-sarenac committed:
Refactor: split class has no references or is empty on two conditions

balsa-sarenac committed:
Fix: use new split preconditinos in remove class refactoring

balsa-sarenac committed:
Fix: add missing description message for BrowseClassReferencesChoice

balsa-sarenac committed:
Feat: configurable dialog for breaking changes

Create a dialog that has choices based on preconditions that failed

balsa-sarenac committed:
Cleanup: white spaces

balsa-sarenac committed:
Refactor: lazy initialization of violators for no subclasses condition

balsa-sarenac committed:
Fix: don't check preconditions twice

Since Driver already checked all precondition we only need changes
so `generateChagnes`

balsa-sarenac committed:
Refactor: lazily accessor for violators to avoid redundant executions

balsa-sarenac committed:
Fix: make sure violators actually return correct value

balsa-sarenac committed:
Feat: initial verison of broswe class references

balsa-sarenac committed:
Fix: protocol names in remove class driver

balsa-sarenac committed:
Feat: initial implementation of remove class pushing state to subclasses

balsa-sarenac committed:
Refactor: remove overriding preconditions (same as superclass)

balsa-sarenac committed:
Fix: return breaking changes to RemoveClassRefactoring

balsa-sarenac committed:
Fix: when pushing state to subclasses, we don't need to check if class is empty

balsa-sarenac committed:
Fix: correctly group preconditions

balsa-sarenac committed:
Fix: pass model to  RemoveClassPushingStateToSubclasses

balsa-sarenac committed:
Test: migrate remove class empty class precondition test to new API

balsa-sarenac committed:
Feat: show message depending on condition

balsa-sarenac committed:
Refactor: rename `empty` to `emptyClasses`

balsa-sarenac committed:
Refactor: extend condition API to improve readability and encapsulation

balsa-sarenac committed:
Hack: add error blocks to breaking change conditions

This is required so that they trigger warning, otherwise they raise error...

balsa-sarenac committed:
Refactor: remove `errorBlockFor:` from `RBAbstractCondition` hierarchy

This method was used for Conjuctive and Negation to give correct
error block OR return nil if the result is true.
This method is not used in the system....
Also when we call errorBlock we _always_ have failed checks...
Removing this method enabled polymorphism for `errorBlock` retrieval.

balsa-sarenac committed:
Test: fix remaining failing tests for RemoveClassRefactoring

Ducasse committed:
Merge pull request #2 from balsa-sarenac/2023-09-16-RestartRemoveClass

Dynamically created dialog with breaking change options

balsa-sarenac committed:
Fix: remove `ifFalse:` from condition

balsa-sarenac committed:
Fix: browse references was always shown

Ducasse committed:
Merge pull request #3 from balsa-sarenac/2023-09-16-RestartRemoveClass

Fixes for remove class dialog

balsa-sarenac committed:
Feat: update descriptions based on condition values

balsa-sarenac committed:
Feat: if there is only one action perform it

This saves user a click, and every click is expensive :)

balsa-sarenac committed:
Cleanup: update spelling in descriptions

balsa-sarenac committed:
Refactor: use `String streamContents:`

jecisc commented:
SycRemoveClassCommandRB2 SycRemoveMethod2Command have unused variables
RBRemoveClassAndPushStateToSubclassChoice RBRemoveClassReparentChoice have uncategorized methods

The other failing tests should be resolved by merge Pharo12 into this branch I think

balsa-sarenac committed:
Refactor: update violators to have more consistent data structures

balsa-sarenac committed:
Refactor: don't scan whole system for class references

Use already calculted references that respect the scope.

balsa-sarenac committed:
Refactor: class references violators are only classes that have references

Previously it was all classes, so we needed to check which ones
have empty collections for references.
This broke encapsulation and the code had to be duplicated.

Ducasse committed:
Merge pull request #4 from balsa-sarenac/2023-09-16-RestartRemoveClass

Small improvements and fixes

Ducasse commented:
Cyril we will fix the integration like the unused var later because this PR is starting to get long.
